crates/executor/src/blockchain_tree/chain.rs

```diff
@@ -0,0 +1,319 @@
+    execution_result::ExecutionResult,
+    substate::{SubStateData, SubStateWithProvider},
+};
+use reth_interfaces::{consensus::Consensus, executor::Error as ExecError, Error};
+use reth_primitives::{BlockHash, BlockNumber, SealedBlockWithSenders, SealedHeader, U256};
+use reth_provider::{BlockExecutor, ExecutorFactory, StateProvider};
+use std::collections::BTreeMap;
+
+/// Side chain that contain it state and connect to block found in canonical chain.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct Chain {
+    /// Chain substate
+    pub substate: SubStateData,
+    /// Changesets for block and transaction.
+    pub changesets: Vec<ExecutionResult>,
+    /// Blocks in this chain
+    pub blocks: BTreeMap<BlockNumber, SealedBlockWithSenders>,
+}
+
+/// Where does the chain connect to.
+#[derive(Clone, Copy)]
+pub struct ForkBlock {
+    /// Block number of block that chains branches from
+    pub number: u64,
+    /// Block hash of block that chains branches from
+    pub hash: BlockHash,
+}
+
+/// Chain identification
+pub type ChainId = u64;
+
+impl Chain {
+    /// Return fork block number and hash.
+    pub fn fork_block(&self) -> ForkBlock {
+        let tip = self.first();
+        ForkBlock { number: tip.number - 1, hash: tip.parent_hash }
+    }
+
+    /// Block fork number
+    pub fn fork_block_number(&self) -> BlockNumber {
+        self.first().number - 1
+    }
+
+    /// Block fork hash
+    pub fn fork_block_hash(&self) -> BlockHash {
+        self.first().parent_hash
+    }
+
+    /// First block in chain.
+    pub fn first(&self) -> &SealedBlockWithSenders {
+        self.blocks.first_key_value().expect("Chain has at least one block for first").1
+    }
+
+    /// Return tip of the chain. Chain always have at least one block inside
+    pub fn tip(&self) -> &SealedBlockWithSenders {
+        self.last()
+    }
+
+    /// Return tip of the chain. Chain always have at least one block inside
+    pub fn last(&self) -> &SealedBlockWithSenders {
+        self.blocks.last_key_value().expect("Chain has at least one block for last").1
+    }
+
+    /// Create new chain with given blocks and execution result.
+    pub fn new(blocks: Vec<(SealedBlockWithSenders, ExecutionResult)>) -> Self {
+        let (blocks, changesets): (Vec<_>, Vec<_>) = blocks.into_iter().unzip();
+
+        let blocks = blocks.into_iter().map(|b| (b.number, b)).collect::<BTreeMap<_, _>>();
+
+        let mut substate = SubStateData::default();
+        substate.apply(&changesets);
+
+        Self { substate, changesets, blocks }
+    }
+
+    /// Create new chain that joins canonical block
+    /// If parent block is the tip mark chain fork.
+    pub fn new_canonical_fork<SP: StateProvider, C: Consensus, EF: ExecutorFactory>(
+        block: &SealedBlockWithSenders,
+        parent_header: SealedHeader,
+        canonical_block_hashes: &BTreeMap<BlockNumber, BlockHash>,
+        provider: &SP,
+        consensus: &C,
+        factory: &EF,
```

Comment:
I am on the fence here, we could have multiple chains so I look at them more like data structures, if consensus/factory are moved inside chain it will be in two places both in tree and chain. I would like to leave it as it is.